src/language-js/print/assignment.js

```diff
@@ -230,41 +242,86 @@ function isAssignmentOrVariableDeclarator(node) {
   return isAssignment(node) || node.type === "VariableDeclarator";
 }
 
-function isMemberExpressionChainHead(node) {
-  return node.type === "Identifier" || node.type === "ThisExpression";
-}
+/**
+ * A chain with no calls at all or whose calls are all without arguments or with lone short arguments,
+ * excluding chains printed by `printMemberChain`
+ */
+function isPoorlyBreakableMemberOrCallChain(
+  path,
+  options,
+  print,
+  deep = false
+) {
+  const node = path.getValue();
+  const goDeeper = () =>
+    isPoorlyBreakableMemberOrCallChain(path, options, print, true);
 
-function isMemberExpressionChainWithSimpleCalls(node) {
-  if (!isMemberExpression(node)) {
-    return false;
+  if (node.type === "TSNonNullExpression") {
+    return path.call(goDeeper, "expression");
   }
-  let { object } = node;
-  for (;;) {
-    if (object.type === "TSNonNullExpression") {
-      object = object.expression;
-    } else if (isCallExpression(object)) {
-      if (!isSimpleCall(object)) {
-        return false;
-      }
-      object = object.callee;
-    } else {
-      break;
+
+  if (isCallExpression(node)) {
+    /** @type {any} TODO */
+    const doc = printCallExpression(path, options, print);
+    if (doc.label === "member-chain") {
+      return false;
+    }
+
+    const args = getCallArguments(node);
+    const isPoorlyBreakableCall =
+      args.length === 0 ||
+      (args.length === 1 && isLoneShortArgument(args[0], options));
+    if (!isPoorlyBreakableCall) {
+      return false;
     }
+
+    return path.call(goDeeper, "callee");
   }
-  return (
-    isMemberExpressionChainHead(object) ||
-    isMemberExpressionChainWithSimpleCalls(object)
-  );
+
+  if (isMemberExpression(node)) {
+    return path.call(goDeeper, "object");
+  }
+
+  return deep && (node.type === "Identifier" || node.type === "ThisExpression");
 }
 
-function isSimpleCall(node) {
-  if (!isCallExpression(node)) {
+const LONE_SHORT_ARGUMENT_THRESHOLD_RATE = 0.25;
+
+function isLoneShortArgument(node, { printWidth }) {
+  const threshold = printWidth * LONE_SHORT_ARGUMENT_THRESHOLD_RATE;
+
+  if (hasComment(node)) {
     return false;
   }
-  const args = getCallArguments(node);
-  return (
-    args.length === 0 || (args.length === 1 && isSimpleCallArgument(args[0], 1))
-  );
+
+  if (
+    node.type === "ThisExpression" ||
```

Comment:
I don't really understand what this function do, but I  guess it's checking short stuff, so `super`? `import.meta`? `new.target`?

Reply:
Or `await short`?

Reply:
```js
const bifornCringerMoshedPerplexSawderGlyphsHa =
  someBigFunctionName("foo")("bar");
```
If each call in RHS is either without arguments or has a single argument, for which this function returns `true`, then breaking after `=` has priority over breaking inside RHS. This way we decrease the likelihood of formatting like this:
```js
const bifornCringerMoshedPerplexSawderGlyphsHa = someBigFunctionName(
  "foo"
)("bar");
```
Yes, there is room for improvement in this function, but it already covers the most important cases and looks good enough. At least for the projects we are testing Prettier on. Lone numbers, short strings, and short identifiers look the worst. It can be polished it later

Reply:
#10222 included similar logic. That's why this specific example is already formatted the preferred way. But the heuristic in #10222 reused `isSimpleCallArgument` from the member chain breaking heuristic, which wasn't exactly a good fit for this case and lead to diffs like the one shown in the [first comment](https://github.com/prettier/prettier/pull/10672#issue-612985282) here.

Reply:
As for `await short`, `import.meta`,  `new.target`, for now I decided to match only one-word things. `super` can't be passed as a call argument.

Reply:
Right.